Update parallel computing docs regarding configuration (#1221)
nest/nest-simulator: doc/guides/parallel_computing.rst

Guide to parallel computing
==============================

.. contents::
   :local:

What is parallelization?
--------------------------

Parallelization can improve the efficiency of running large-scale simulations by
taking advantage of multicore/multiprocessor machines, computer clusters or
supercomputers. Here we explain how  parallelization is set up in NEST and how you
can take advantage of it for your simulations.

NEST employs two methods for parallelization:

* :ref:`Thread-parallel simulation <multiple_threads>`
     * uses OpenMP
     * takes advantage of multicore and multiprocessor computers without
       the need for additional libraries
* :ref:`Distributed simulation (or distributed computing) <distributed_computing>`
     * uses the Message Passing Interface (MPI)
     * supports simulations over multiple computers

Both methods can be combined within a simulation.

See `Plesser et al. (2007) <http://dx.doi.org/10.1007/978-3-540-74466-5_71>`__
for more information on NEST parallelization and be sure to check the

> documentation on `Random numbers in NEST <random-numbers.md>`__.

Reviewer: ```suggestion
documentation on :ref:`Random numbers in NEST <random_numbers>`.
```
Then please add ``.. _random_numbers:`` at the top of ``random_numbers.rst``.

Virtual processes are distributed round-robin (i.e. each VP is allocated equal
time slices, without any given a priority) onto the MPI processes and
counted continuously over all processes.

.. figure:: ../_static/img/Process_vp_thread.png

 Basic scheme showing how threads (T) and virtual
 processes (VP) reside in MPI processes (P) in NEST

Node distributions
--------------------

The distribution of nodes depends on the type of node.

In the figure below, a node distribution for a small network consisting of ``spike_generator``,
four ``iaf_psc_alpha`` neurons, and a ``spike_detector``
in a scenario with two processes with two threads each.

.. figure:: ../_static/img/Node_distribution.png

 sg=spike\_generator, iaf=iaf\_psc\_alpha, sd=spike\_detector. Numbers to
 the left and right indicate global ids.
 The *proxy* object in the figure is a conceptual way of keeping the id of the
 real node free on remote processes).

Neuron distribution
~~~~~~~~~~~~~~~~~~~~

Neurons are assigned to one of the virtual processes in a round-robin fashion.
On all other virtual processes, no object is created. Proxies ensure the id
of the real node on a given VP is kept free.

> The virtual process `idVP` on which a neuron with global id `idNode` is

Reviewer: ```suggestion
The virtual process :math:`id_{vp}` on which a neuron with global id :math:`gid_{node}` is
```

> allocated is given by :math:`idVP = idNode %NVP`, where `NVP` is the total

Reviewer: ```suggestion
allocated is given by :math:`id_{vp} = gid_{node} %N_{vp}`, where :math:`N_{vp}` is the total
```

number of virtual processes in the simulation.

Device Distribution
~~~~~~~~~~~~~~~~~~~~~~~

Devices are replicated once on each thread in order to balance the load and
minimize their interaction. Devices thus do not have proxies on remote virtual
processes.

For recording devices configured to record to a file (property
`to_file` set to `true`), the distribution results in multiple
data files, each containing the data from one thread. The files names
are composed according to the following scheme

.. code-block:: bash

    [model|label]-gid-vp.[dat|gdf]

The first part is the name of the `model` (e.g. ``voltmeter`` or
``spike_detector``) or, if set, the `label` of the recording device. Next is
the global id (GID) of the recording device, followed by the id of the VP
assigned to the recorder. Spike files have the file extension ``gdf`` and
analog recordings from the ``multimeter`` have ``dat`` as file extension.

The ``label`` and ``file_extension`` of a recording device can be set like any
other parameter of a node using ``SetStatus``.

> .. note::

Reviewer: Please move this note between Figure 13 and "Neuron distribution"

 The status dictionary of each node (i.e. neuron or device) contains
 three entries that are related to parallel computing:

 *  *local* (boolean): indicating if the node exists on the local process or not
 *  *thread* (integer): id of the local thread the node is assigned to

>  *  *vp* (integer): id of the virtual process the node is assigned to

Reviewer: ```suggestion
 *  *vp* (integer): id of the virtual process the node is assigned to (denoted :math:`id_{vp}` here)
```
